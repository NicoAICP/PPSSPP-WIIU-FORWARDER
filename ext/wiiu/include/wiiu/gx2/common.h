#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
