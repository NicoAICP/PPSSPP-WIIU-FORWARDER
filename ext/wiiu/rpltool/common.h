#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
