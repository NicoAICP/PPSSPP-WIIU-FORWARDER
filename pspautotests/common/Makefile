TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
