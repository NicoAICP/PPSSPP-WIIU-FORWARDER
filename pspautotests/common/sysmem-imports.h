extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();