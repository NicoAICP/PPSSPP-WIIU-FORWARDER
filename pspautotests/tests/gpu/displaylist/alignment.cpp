#include <common.h>

#include <stdlib.h>
#include <stdio.h>
#include <pspgu.h>
#include <pspgum.h>
#include <pspthreadman.h>
#include <pspkernel.h>

extern "C" {
#include "sysmem-imports.h"
#include "../commands/commands.h"
}

u32 dlistReset[] = {
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000000,
	// Try to make sure we always finish.
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
};

u32 dlist1[] = {
	// Add a NOP that looks like an instruction when read at an offset.
	(GE_CMD_NOP << 24) | (GE_CMD_AMBIENTCOLOR << 16) | (GE_CMD_AMBIENTCOLOR << 8) | GE_CMD_AMBIENTCOLOR,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000001,
	// Try to make sure we always finish.
	(GE_CMD_FINISH << 24) | (GE_CMD_FINISH << 16) | (GE_CMD_FINISH << 8) | GE_CMD_FINISH,
	(GE_CMD_END << 24) | (GE_CMD_END < 16) | (GE_CMD_END << 8) | GE_CMD_END,
	(GE_CMD_NOP << 24),
};

u32 dlist2[] = {
	(GE_CMD_BASE << 24) | 0x000000,
	(GE_CMD_ORIGIN << 24),
	// Didn't jump at all?
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000000,
	// This is overwritten by the actual jump offset.
	(GE_CMD_JUMP << 24) | 0x00001D,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000009,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,

	// Add a NOP that looks like an instruction when read at an offset.
	(GE_CMD_NOP << 24) | (GE_CMD_AMBIENTCOLOR << 16) | (GE_CMD_AMBIENTCOLOR << 8) | GE_CMD_AMBIENTCOLOR,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000002,
	// Try to make sure we always finish.
	(GE_CMD_FINISH << 24) | (GE_CMD_FINISH << 16) | (GE_CMD_FINISH << 8) | GE_CMD_FINISH,
	(GE_CMD_END << 24) | (GE_CMD_END < 16) | (GE_CMD_END << 8) | GE_CMD_END,
	(GE_CMD_NOP << 24),
};

u32 dlist3[] = {
	(GE_CMD_BOUNDINGBOX << 24) | 0x000000,
	(GE_CMD_BASE << 24) | 0x000000,
	(GE_CMD_ORIGIN << 24),
	// Didn't jump at all?
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000000,
	// This is overwritten by the actual jump offset.
	(GE_CMD_BJUMP << 24) | (0x00001D),
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000009,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,

	// Add a NOP that looks like an instruction when read at an offset.
	(GE_CMD_NOP << 24) | (GE_CMD_AMBIENTCOLOR << 16) | (GE_CMD_AMBIENTCOLOR << 8) | GE_CMD_AMBIENTCOLOR,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000003,
	// Try to make sure we always finish.
	(GE_CMD_FINISH << 24) | (GE_CMD_FINISH << 16) | (GE_CMD_FINISH << 8) | GE_CMD_FINISH,
	(GE_CMD_END << 24) | (GE_CMD_END < 16) | (GE_CMD_END << 8) | GE_CMD_END,
	(GE_CMD_NOP << 24),
};

u32 dlist4[] = {
	(GE_CMD_BASE << 24) | 0x000000,
	(GE_CMD_ORIGIN << 24),
	// Didn't jump at all?
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000000,
	// This is overwritten by the actual jump offset.
	(GE_CMD_CALL << 24) | 0x00001D,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,

	// Add a NOP that looks like an instruction when read at an offset.
	(GE_CMD_NOP << 24) | (GE_CMD_AMBIENTCOLOR << 16) | (GE_CMD_AMBIENTCOLOR << 8) | GE_CMD_AMBIENTCOLOR,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000004,
	// Try to make sure we always finish.
	(GE_CMD_RET << 24) | (GE_CMD_RET < 16) | (GE_CMD_RET << 8) | GE_CMD_RET,
	(GE_CMD_NOP << 24),
};

u32 dlist5[] = {
	(GE_CMD_BASE << 24) | 0x000000,
	(GE_CMD_ORIGIN << 24),
	// Didn't jump at all?
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000000,
	// These two are overwritten by the actual jump offset.
	(GE_CMD_SIGNAL << 24) | (0x10 << 16) | 0x0000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000009,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,

	// Add a NOP that looks like an instruction when read at an offset.
	(GE_CMD_NOP << 24) | (GE_CMD_AMBIENTCOLOR << 16) | (GE_CMD_AMBIENTCOLOR << 8) | GE_CMD_AMBIENTCOLOR,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000005,
	// Try to make sure we always finish.
	(GE_CMD_FINISH << 24) | (GE_CMD_FINISH << 16) | (GE_CMD_FINISH << 8) | GE_CMD_FINISH,
	(GE_CMD_END << 24) | (GE_CMD_END < 16) | (GE_CMD_END << 8) | GE_CMD_END,
	(GE_CMD_NOP << 24),
};

u32 dlist6[] = {
	(GE_CMD_BASE << 24) | 0x000000,
	(GE_CMD_ORIGIN << 24),
	// Didn't jump at all?
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000000,
	// These two are overwritten by the actual jump offset.
	(GE_CMD_SIGNAL << 24) | (0x11 << 16) | 0x0000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,

	// Add a NOP that looks like an instruction when read at an offset.
	(GE_CMD_NOP << 24) | (GE_CMD_AMBIENTCOLOR << 16) | (GE_CMD_AMBIENTCOLOR << 8) | GE_CMD_AMBIENTCOLOR,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000006,
	// This is the signal return.
	(GE_CMD_SIGNAL << 24) | (0x12 << 16) | 0x0000,
	(GE_CMD_END << 24) | 0x000000,
	// Try to make sure we always finish, just to be safe.
	(GE_CMD_FINISH << 24) | (GE_CMD_FINISH << 16) | (GE_CMD_FINISH << 8) | GE_CMD_FINISH,
	(GE_CMD_END << 24) | (GE_CMD_END < 16) | (GE_CMD_END << 8) | GE_CMD_END,
	(GE_CMD_NOP << 24),
};

u32 runList(void *list, void *listend, bool head = false) {
	int listid = head ? sceGeListEnQueueHead(list, list, -1, NULL) : sceGeListEnQueue(list, list, -1, NULL);
	if (listid < 0) {
		return listid;
	}
	sceGeListUpdateStallAddr(listid, listend);
	if (head) {
		sceGeContinue();
	}
	sceKernelDelayThread(5000);
	sceGeBreak(1, NULL);

	return sceGeGetCmd(GE_CMD_AMBIENTCOLOR);
}

u32 runListAndReset(void *list, void *listend, bool head = false) {
	u32 result = runList(list, listend, head);
	runList(dlistReset, dlistReset + 100);
	return result;
}

u32 runJumpList(int offset) {
	dlist2[3] = (GE_CMD_JUMP << 24) | (0x000018 + offset);
	sceKernelDcacheWritebackInvalidateRange(dlist2, sizeof(dlist2));

	return runList(dlist2, dlist2 + 100);
}

u32 runBJumpList(int offset) {
	dlist3[4] = (GE_CMD_BJUMP << 24) | (0x000018 + offset);
	sceKernelDcacheWritebackInvalidateRange(dlist3, sizeof(dlist3));

	return runList(dlist3, dlist3 + 100);
}

u32 runCallList(int offset) {
	dlist4[3] = (GE_CMD_CALL << 24) | (0x000014 + offset);
	sceKernelDcacheWritebackInvalidateRange(dlist4, sizeof(dlist4));

	return runList(dlist4, dlist4 + 100);
}

u32 runSignalJumpList(int offset) {
	u32 target = (intptr_t)dlist5 + 0x20 + offset;
	dlist5[3] = (GE_CMD_SIGNAL << 24) | (0x10 << 16) | (target >> 16);
	dlist5[4] = (GE_CMD_END << 24) | (target & 0xFFFF);
	sceKernelDcacheWritebackInvalidateRange(dlist5, sizeof(dlist5));

	return runList(dlist5, dlist5 + 100);
}

u32 runSignalCallList(int offset) {
	u32 target = (intptr_t)dlist6 + 0x1C + offset;
	dlist6[3] = (GE_CMD_SIGNAL << 24) | (0x10 << 16) | (target >> 16);
	dlist6[4] = (GE_CMD_END << 24) | (target & 0xFFFF);
	sceKernelDcacheWritebackInvalidateRange(dlist6, sizeof(dlist6));

	return runList(dlist6, dlist6 + 100);
}

extern "C" int main(int argc, char *argv[]) {
	checkpointNext("Enqueue");
	checkpoint("  Aligned: %x", runListAndReset(dlist1, dlist1 + 100));
	checkpoint("  Offset + 1: %x", runListAndReset((u8 *)dlist1 + 1, dlist1 + 100));
	checkpoint("  Offset + 2: %x", runListAndReset((u8 *)dlist1 + 2, dlist1 + 100));
	checkpoint("  Offset + 3: %x", runListAndReset((u8 *)dlist1 + 3, dlist1 + 100));
	checkpoint("  Offset + 4: %x", runListAndReset((u8 *)dlist1 + 4, dlist1 + 100));
	checkpoint("  Offset + 8: %x", runListAndReset((u8 *)dlist1 + 8, dlist1 + 100));

	checkpointNext("EnqueueHead");
	checkpoint("  Aligned: %x", runListAndReset(dlist1, dlist1 + 100, true));
	checkpoint("  Offset + 1: %x", runListAndReset((u8 *)dlist1 + 1, dlist1 + 100, true));
	checkpoint("  Offset + 2: %x", runListAndReset((u8 *)dlist1 + 2, dlist1 + 100, true));
	checkpoint("  Offset + 3: %x", runListAndReset((u8 *)dlist1 + 3, dlist1 + 100, true));
	checkpoint("  Offset + 4: %x", runListAndReset((u8 *)dlist1 + 4, dlist1 + 100, true));
	checkpoint("  Offset + 8: %x", runListAndReset((u8 *)dlist1 + 8, dlist1 + 100, true));

	checkpointNext("Jump");
	checkpoint("  Aligned: %x", runJumpList(0));
	checkpoint("  Offset + 1: %x", runJumpList(1));
	checkpoint("  Offset + 2: %x", runJumpList(2));
	checkpoint("  Offset + 3: %x", runJumpList(3));
	checkpoint("  Offset + 4: %x", runJumpList(4));
	checkpoint("  Offset + 8: %x", runJumpList(8));

	checkpointNext("BJump");
	checkpoint("  Aligned: %x", runBJumpList(0));
	checkpoint("  Offset + 1: %x", runBJumpList(1));
	checkpoint("  Offset + 2: %x", runBJumpList(2));
	checkpoint("  Offset + 3: %x", runBJumpList(3));
	checkpoint("  Offset + 4: %x", runBJumpList(4));
	checkpoint("  Offset + 8: %x", runBJumpList(8));

	checkpointNext("Call");
	checkpoint("  Aligned: %x", runCallList(0));
	checkpoint("  Offset + 1: %x", runCallList(1));
	checkpoint("  Offset + 2: %x", runCallList(2));
	checkpoint("  Offset + 3: %x", runCallList(3));
	checkpoint("  Offset + 4: %x", runCallList(4));
	checkpoint("  Offset + 8: %x", runCallList(8));

	checkpointNext("Signal Jump");
	checkpoint("  Aligned: %x", runSignalJumpList(0));
	checkpoint("  Offset + 1: %x", runSignalJumpList(1));
	checkpoint("  Offset + 2: %x", runSignalJumpList(2));
	checkpoint("  Offset + 3: %x", runSignalJumpList(3));
	checkpoint("  Offset + 4: %x", runSignalJumpList(4));
	checkpoint("  Offset + 8: %x", runSignalJumpList(8));

	checkpointNext("Signal Call");
	checkpoint("  Aligned: %x", runSignalCallList(0));
	checkpoint("  Offset + 1: %x", runSignalCallList(1));
	checkpoint("  Offset + 2: %x", runSignalCallList(2));
	checkpoint("  Offset + 3: %x", runSignalCallList(3));
	checkpoint("  Offset + 4: %x", runSignalCallList(4));
	checkpoint("  Offset + 8: %x", runSignalCallList(8));

	return 0;
}