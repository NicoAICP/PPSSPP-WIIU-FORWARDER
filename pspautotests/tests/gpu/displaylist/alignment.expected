[x] Enqueue
[r]   Aligned: 5c000001
[r]   Offset + 1: 80000103
[r]   Offset + 2: 80000103
[r]   Offset + 3: 80000103
[r]   Offset + 4: 5c000001
[r]   Offset + 8: 5c000000

[x] EnqueueHead
[r]   Aligned: 5c000001
[r]   Offset + 1: 80000103
[r]   Offset + 2: 80000103
[r]   Offset + 3: 80000103
[r]   Offset + 4: 5c000001
[r]   Offset + 8: 5c000000

[x] Jump
[r]   Aligned: 5c000002
[r]   Offset + 1: 5c000002
[r]   Offset + 2: 5c000002
[r]   Offset + 3: 5c000002
[r]   Offset + 4: 5c000002
[r]   Offset + 8: 5c000000

[x] BJump
[r]   Aligned: 5c000003
[r]   Offset + 1: 5c000003
[r]   Offset + 2: 5c000003
[r]   Offset + 3: 5c000003
[r]   Offset + 4: 5c000003
[r]   Offset + 8: 5c000000

[x] Call
[r]   Aligned: 5c000004
[r]   Offset + 1: 5c000004
[r]   Offset + 2: 5c000004
[r]   Offset + 3: 5c000004
[r]   Offset + 4: 5c000004
[r]   Offset + 8: 5c000000

[x] Signal Jump
[r]   Aligned: 5c000005
[r]   Offset + 1: 5c000005
[r]   Offset + 2: 5c000005
[r]   Offset + 3: 5c000005
[r]   Offset + 4: 5c000005
[r]   Offset + 8: 5c000000

[x] Signal Call
[r]   Aligned: 5c000006
[r]   Offset + 1: 5c000006
[r]   Offset + 2: 5c000006
[r]   Offset + 3: 5c000006
[r]   Offset + 4: 5c000006
[r]   Offset + 8: 5c000000
